pkg/storaged/pages.jsx

```diff
@@ -0,0 +1,691 @@
+    if (!card)
+        return null;
+    return location_from_card(card.next) || card.page_location;
+}
+
+function size_from_card(card) {
+    if (!card)
+        return null;
+    if (card.page_size)
+        return card.page_size;
+    return size_from_card(card.next);
+}
+
+export function new_page(parent, card, options) {
+    const page = {
+        location: location_from_card(card),
+        name: name_from_card(card),
+        icon: icon_from_card(card),
+        key: key_from_card(card),
+        parent,
+        children: [],
+        card,
+        options: options || {},
+    };
+    page.columns = [
+        card.title,
+        card.location,
+        size_from_card(card),
+    ];
+    if (parent)
+        parent.children.push(page);
+    while (card) {
+        card.page = page;
+        card = card.next;
+    }
+    if (page.location) {
+        pages.set(JSON.stringify(page.location), page);
+        if (page.location.length == 0) {
+            // This is the Overview page. Make it the parent of the
+            // special "not found" page (but don't make the "not
+            // found" page a child of the Overview...)
+            not_found_page.parent = page;
+        }
+    }
+    return page;
+}
+
+export function new_card({
+    title, location, next,
+    type_extra, id_extra,
+    page_name, page_icon, page_key, page_location, page_size,
+    page_block,
+    for_summary,
+    has_warning, has_danger, job_path,
+    component, props,
+    actions,
+}) {
+    if (page_block) {
+        page_location = [block_location(page_block)];
+        page_name = block_short_name(page_block);
+        page_size = page_block.Size;
+        job_path = page_block.path;
+    }
+    return {
+        title,
+        location,
+        next,
+        type_extra,
+        page_name,
+        page_icon,
+        page_key,
+        page_location,
+        page_size,
+        for_summary,
+        component,
+        props,
+        has_warning,
+        has_danger,
+        job_path,
+        actions: actions ? actions.filter(a => !!a) : null,
+        id_extra,
+    };
+}
+
+export function register_crossref(crossref) {
+    const val = crossrefs.get(crossref.key) || [];
+    crossref.actions = crossref.actions ? crossref.actions.filter(a => !!a) : null;
+    val.push(crossref);
+    crossrefs.set(crossref.key, val);
+}
+
+export function get_crossrefs(key) {
+    return crossrefs.get(key);
+}
+
+/* Getting the page for a navigation location.
+ *
+ * We have a special "not found" page that is returned when there is
+ * no real page at the given location.
+ */
+
+const NotFoundCard = ({ card }) => {
+    return <span>{_("Not found")}</span>;
+};
+
+const not_found_page = new_page(null, new_card({ page_name: _("Not found"), component: NotFoundCard }));
+
+export function get_page_from_location(location) {
+    if (!pages)
+        return not_found_page;
+
+    return pages.get(JSON.stringify(location)) || not_found_page;
```

Comment:
I promised myself to not make JS syntax comments, but for these cases `pages?.get() || not_found_page` is really a helpful construct to avoid the special case.

Reply:
Done!